controller/consul_client_token.go

```diff
@@ -0,0 +1,93 @@
+package controller
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/aws/aws-sdk-go/service/secretsmanager/secretsmanageriface"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+)
+
+// UpsertConsulClientToken creates or updates ACL policy and token for the Consul client in Consul.
+// It then saves the created token in AWS Secrets Manager in the secret provided by secretARN.
+func UpsertConsulClientToken(consulClient *api.Client, smClient secretsmanageriface.SecretsManagerAPI, secretARN, secretPrefix string, log hclog.Logger) error {
+	// Read the secret from AWS.
+	currSecretValue, err := smClient.GetSecretValue(&secretsmanager.GetSecretValueInput{
+		SecretId: aws.String(secretARN),
+	})
+	if err != nil {
+		return fmt.Errorf("retrieving Consul client secret: %w", err)
+	}
+
+	// Unmarshal the secret value JSON.
+	var currSecret tokenSecretJSON
+	err = json.Unmarshal([]byte(*currSecretValue.SecretString), &currSecret)
+	if err != nil {
+		return fmt.Errorf("unmarshalling Consul client secret value JSON: %w", err)
+	}
+
+	var currToken *api.ACLToken
+
+	// If the secret is not empty, check if Consul already has a token with this AccessorID.
+	if currSecret.AccessorID != "" {
+		currToken, _, err = consulClient.ACL().TokenRead(currSecret.AccessorID, nil)
+		if err != nil && !isACLNotFoundError(err) {
+			return fmt.Errorf("reading token: %w", err)
+		}
+	}
+
+	// If there is no current token in Consul, we need to create one.
+	if currToken == nil {
+		// First, we need to check if the policy for the Consul client already exists.
+		// If it does, we will skip policy creation.
+		policyName := fmt.Sprintf("%s-consul-client-policy", secretPrefix)
+		policy, _, err := consulClient.ACL().PolicyReadByName(policyName, nil)
+
+		// When policy is not found, Consul returns ACL not found error.
+		if isACLNotFoundError(err) {
+			// Create a policy for the Consul clients.
+			log.Info("creating ACL policy", "name", policyName)
+			policy, _, err = consulClient.ACL().PolicyCreate(&api.ACLPolicy{
+				Name:        policyName,
+				Description: "Consul Client Token Policy for ECS",
+				Rules:       `node_prefix "" { policy = "write" } service_prefix "" { policy = "read" }`,
```

Comment:
worth a note that all clients share a token so that's why we're using prefix empty?